### DISCUSSION TITLE
Please confirm my understanding of RemoteBucketBuilder.build(K, Supplier)

### DISCUSSION THREAD

**Discussion by sp00m** [vladimir-bukhtoyarov/bucket4j#492] [Q&A] [answered]
Hello,

I used to create bucket proxies backed by Redis as follows:

```java
var bucketProxy = proxyManager.builder().build(name, config);
```

While upgrading our dependencies, I noticed `#build` went deprecated in favour of the Supplier-based method:

```
var bucketProxy = proxyManager.builder().build(name, () -> config);
```

Its javadoc states:

> [The Supplier] will be called if and only if bucket has not been persisted before.

Also, from the javadoc of `ExpirationAfterWriteStrategy#basedOnTimeForRefillingBucketUpToMax`:

> [The argument] specifies how long bucket should be held in the cache after all consumed tokens have been refilled.

Does it mean that, if for some reason I need to update the configured limits (e.g. from 5 operations by second to 10), it will only be applied once the existing 5-operation bucket has been refilled (plus that custom duration), because in this case, the bucket will be deleted from the cache, so the config Supplier will need to be called again? Perhaps this behaviour was still there before but has been made more obvious with this Supplier-based method.

Thank you 🙌

**Comment by vladimir-bukhtoyarov** [comment, marked as answer]
Hello @sp00m 

Current javadoc for config is little bit inncorect. It should be changed like this: `[The Supplier] will be called if and only if it detected that bucket is not persisted by any reason like was never persisted before, persisted but then expired, persisted but then explicitly removed, or persisted but lost because of any storage problem.`

The words above have following outcome according to configuration replacement: if you plan to use configuration replacement then your config supplier should return as fresh configuration as it possible, otherwise there is risk to lose the config change when bucket with changed configuration bacomes expired and then becomes ressurected with legacy config, becuse config suppliers returns  legacy config.
